sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -454,7 +454,9 @@ public ProcessContinuation processElement(
       final Stopwatch sw = Stopwatch.createStarted();
 
       while (true) {
-        rawRecords = poll(consumer, kafkaSourceDescriptor.getTopicPartition());
+        KafkaMetrics kafkaMetrics = KafkaSinkMetrics.kafkaMetrics();
+        rawRecords = poll(consumer, kafkaSourceDescriptor.getTopicPartition(), kafkaMetrics);
+        kafkaMetrics.flushBufferedMetrics();
```

Comment:
what about not flushing here and just falling through?

would it be ok to just flush once outside the while loop with a try { } finally { kafkaMetrics.flushBufferedMetrics(); } block?

Reply:
I didn't do it because of the returns, but apparently the finally block will run and then return, so that will work. I have modified the code.